ca/ca.go

```diff
@@ -179,25 +179,59 @@ func (ca *CAImpl) newIntermediateIssuer(root *issuer, intermediateKey crypto.Sig
 	if err != nil {
 		return nil, err
 	}
-	ca.log.Printf("Generated new intermediate issuer with serial %s and SKI %x\n", ic.ID, subjectKeyID)
+	ca.log.Printf("Generated new intermediate issuer %s with serial %s and SKI %x\n", ic.Cert.Subject, ic.ID, subjectKeyID)
 	return &issuer{
 		key:  intermediateKey,
 		cert: ic,
 	}, nil
 }
 
-func (ca *CAImpl) newChain(intermediateKey crypto.Signer, intermediateSubject pkix.Name, subjectKeyID []byte) *chain {
+// newChain generates a new issuance chain, including a root certificate and numIntermediates intermediates (at least 1).
+// The first intermediate will use intermediateKey, intermediateSubject and subjectKeyId.
+// Any intermediates between the first intermediate and the root will have their keys and subjects generated automatically.
+func (ca *CAImpl) newChain(intermediateKey crypto.Signer, intermediateSubject pkix.Name, subjectKeyID []byte, numIntermediates int) *chain {
+	if numIntermediates <= 0 {
+		panic("At least one intermediate must be present in the certificate chain")
+	}
 	root, err := ca.newRootIssuer()
 	if err != nil {
 		panic(fmt.Sprintf("Error creating new root issuer: %s", err.Error()))
 	}
-	intermediate, err := ca.newIntermediateIssuer(root, intermediateKey, intermediateSubject, subjectKeyID)
+
+	// The last N-1 intermediates build a path from the root to the leaf signing certificate.
+	// If numIntermediates is only 1, then no intermediates will be generated here.
+	prev := root
+	intermediates := make([]*issuer, numIntermediates)
+	for i := 0; i < numIntermediates-1; i++ {
+		k, ski, err := makeKey()
+		if err != nil {
+			panic(fmt.Sprintf("Error creating new intermediate issuer: %v", err))
+		}
+		intermediate, err := ca.newIntermediateIssuer(prev, k, pkix.Name{
+			CommonName: intermediateCAPrefix + hex.EncodeToString(makeSerial().Bytes()[:3]),
+		}, ski)
+		if err != nil {
+			panic(fmt.Sprintf("Error creating new intermediate issuer: %s", err.Error()))
+		}
+		intermediates[i] = intermediate
+		prev = intermediate
+	}
+
+	// The first issuer is the one which signs the leaf certificates
+	intermediate, err := ca.newIntermediateIssuer(prev, intermediateKey, intermediateSubject, subjectKeyID)
 	if err != nil {
 		panic(fmt.Sprintf("Error creating new intermediate issuer: %s", err.Error()))
 	}
+	intermediates[numIntermediates-1] = intermediate
+
+	// Reverse the order of the intermediates, from leaf signer to root.
+	for i, j := 0, len(intermediates)-1; i < j; i, j = i+1, j-1 {
```

Comment:
This could be skipped by just reversing the direction of the for loop above, i.e. `for i := numIntermediates-1; i > 0; i--` (when constructing the common name you need to be careful you inverse the index so the numbering makes sense). You'd then stick the last intermediate in `intermediates[0]`.